File: solutions/c/armstrong-numbers/1/armstrong_numbers.c
#include "armstrong_numbers.h"

#include<math.h>
#include<stdio.h>

#define MAX_DIGITS 10

bool is_armstrong_number(int candidate)
{
    int digits[MAX_DIGITS] = {0};
    int n_digits = 0;
    int current = candidate;
    for(int i = 1; current != 0; i++) {
        digits[i] = current % 10;
        n_digits++;
        current /= 10;
    }
    
    if (n_digits == 0) return true;

    int sum = 0;
    for(int i = 0; i < MAX_DIGITS; i++) {
        sum += (int)pow(digits[i], n_digits);
        if (sum > candidate) {
            return false;
        }
    }
    return sum == candidate;
}

File: solutions/ocaml/armstrong-numbers/1/armstrong_numbers.ml
let validate candidate =
  let rec get_digits acc n =
  match n with
    | n when n / 10 = 0 -> n :: acc
    | n -> get_digits ((n mod 10) :: acc) (n / 10)
  in
  let rec pow n p =
    int_of_float (Float.pow (float_of_int n) (float_of_int p))
  in
    let digits = get_digits [] candidate in
    let n_digits = List.length digits in
    let sum = digits
      |> List.map (fun x -> pow x n_digits)
      |> List.fold_left (+) 0 in
    sum = candidate